lxd/db/entity.go

```diff
@@ -0,0 +1,307 @@
+//go:build linux && cgo && !agent
+// +build linux,cgo,!agent
+
+package db
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/lxc/lxd/lxd/db/cluster"
+)
+
+// ErrUnknownEntityID describes the unknown entity ID error
+var ErrUnknownEntityID = fmt.Errorf("Unknown entity ID")
+
+// GetURIFromEntity returns the URI for the given entity type and entity ID.
+func (c *Cluster) GetURIFromEntity(entityType int, entityID int) (string, error) {
```

Comment:
@monstermunchkin in https://github.com/lxc/lxd/pull/8582/commits/2f30dbdba7a151e7d54be5a67f683c46192bb981 you added functions to get objects via their IDs, but this function seems to (mostly) get all objects of a certain type and iterate through them looking for the matching ID. This seems pretty inefficient, especially for instances which will retrieve multiple different queries per instance. Any reason you didnt use the ID oriented retrieval functions you added?

Reply:
Previously, I had added `ID` to the filters, and updated the mapper functions (where possible) to handle that. However, that caused issues in the object retrieval and broke our unit tests. I therefore removed the mapper changes, and moved the filtering to the function itself.